### PR TITLE
JS/Python: Fix FP in redos related to empty lookaheads

### LINKED ISSUE #5964: LGTM.com - false positive javascript/ql/src/Performance/ReDoS.ql
**Description of the false positive**

Happened on
https://github.com/sparkletown/sparkle/pull/1443/files#diff-06ebbfcf8e627f41bd6b15a2648431dc8a5dfd21f7d64f9bc798ed55051c0c58R97
with code

```javascript
.match(/("[^"]*?"|[^"\s]+)+(?=\s*|\s*$)/g) 
```

stating:

    Inefficient regular expression

    This part of the regular expression may cause exponential backtracking on strings
    containing many repetitions of '!'.

although regular expression does not have any `!` and I am not aware of any special handling of `!` in the source string which should somehow trigger such behavior.

**URL to the alert on the project page on LGTM.com**

unfortunately I am just a contributor and do not have access to those. May be @0xdevalias could help.

<details>
<summary>and here is the webshot</summary> 

image
</details>

### PR DESCRIPTION
**Pull request by erik-krogh**
Fixes #5964

--- 

Lookaheads/lookbehinds that match the empty string will always match, and they are thus pretty much useless and unnecessary.  
However, I've now seen multiple regular expressions that have empty lookaheads. 

The ReDoS query assumed that a lookahead had a rejecting suffix, but that assumptions breaks with lookaheads that match the empty string. 

This PR fixes that by adding an epsilon transition across lookaheads/lookbehinds that match the empty string. 

---

[JavaScript evaluation looks fine](https://github.com/dsp-testing/erik-krogh-dca/tree/run/emptyRedos-js-eval/reports)
[Python evaluation looks fine](https://github.com/dsp-testing/erik-krogh-dca/tree/run/emptyRedos__Differences__CustomSuite/reports)

I don't think this needs a change-note.

### CODE REVIEW

**Comment by asgerf** [comment]
LGTM

I think it would be worth adding some test cases with a lookahead that just checks for an anchor, without an alternative that is also empty. That's one of the main cases where `matchesEpsilon` differs from `RegExpTerm.isNullable`.
```js
/(x*)+(?=$)/
/(x*)+(?=$|y)/
/([\s\S]*)+(?=$)/
/([\s\S]*)+(?=$|y)/
```

File: javascript/ql/test/query-tests/Performance/ReDoS/tst.js [hunk 1]
```diff
@@ -370,4 +370,8 @@ var good43 = /("[^"]*?"|[^"\s]+)+(?=\s*|\s*$)/g;
 
 // BAD
 var bad87 = /("[^"]*?"|[^"\s]+)+(?=\s*|\s*$)X/g;
-var bad88 = /("[^"]*?"|[^"\s]+)+(?=X)/g;
\ No newline at end of file
+var bad88 = /("[^"]*?"|[^"\s]+)+(?=X)/g;
+var bad89 = /(x*)+(?=$)/
+var bad90 = /(x*)+(?=$|y)/
+var bad91 = /([\s\S]*)+(?=$)/
+var bad92 = /([\s\S]*)+(?=$|y)/
```
Comment (asgerf) [inline javascript/ql/test/query-tests/Performance/ReDoS/tst.js:377]: I believe the last two ones should be marked as good? I'm not asking for the query to handle them, just that the test is accurate.

Comment (erik-krogh) [reply javascript/ql/test/query-tests/Performance/ReDoS/tst.js]: You're right. I was a little fast there.

**Comment by erik-krogh** [comment]
ping @asgerf, @yoff  
Can I get a review?

**Review by asgerf — APPROVED** [review]

File: python/ql/src/semmle/python/RegexTreeView.qll [hunk 2]
```diff
@@ -836,6 +836,13 @@ class RegExpZeroWidthMatch extends RegExpGroup {
  */
 class RegExpSubPattern extends RegExpZeroWidthMatch {
   RegExpSubPattern() { not re.emptyGroup(start, end) }
+
+  /** Gets the lookahead term. */
+  RegExpTerm getOperand() {
+    result.getRegex() = re and
+    result.getStart() = start + 3 and
+    result.getEnd() = end - 1
+  }
```
Comment (yoff) [inline python/ql/src/semmle/python/RegexTreeView.qll:845]: This is slightly incorrect. A `RegExpZeroWidthMatch` can arise from `re.negativeAssertionGroup(start, end)`, where `end in start+[3,4]`.

I think you want something like:
```suggestion
  /** Gets the lookahead term. */
  RegExpTerm getOperand() {
    exists(int in_start, int in_end | re.groupContents(start, end, in_start, in_end) |
      result.getRegex() = re and
      result.getStart() = in_start and
      result.getEnd() = in_end
    )
  }
```

**Review by yoff — CHANGES_REQUESTED** [review]
Just one thing to address. I also wonder if `matchesEpsilon` would be nicer as a member predicate on `RegExpTerm`, but I will not insist on that for this PR.

**Review by yoff — APPROVED** [review]
LGTM
